Auto-expand “details” element for find-in-page and scrolling to fragments

sideshowbarker opened the pull request:
#### 2118f670fe3b5d958686ed7f572dd24ce697f6c3
<pre>
Auto-expand “details” element for find-in-page and scrolling to fragments
<a href="https://bugs.webkit.org/show_bug.cgi?id=228843">https://bugs.webkit.org/show_bug.cgi?id=228843</a>
<a href="https://rdar.apple.com/81856620">rdar://81856620</a>

Reviewed by NOBODY (OOPS!).

This change causes any closed &lt;details&gt; element to be opened
automatically (auto-expanded) if either:

- the user is using find-in-page and a descendant node of the closed
  &lt;details&gt; element contains a match for the given search string
- the user is navigating to a fragment ID and the fragment ID is for a
  descendant element of the closed &lt;details&gt; element

…per the requirements in the HTML spec’s Find-in-page section at
<a href="https://html.spec.whatwg.org/#interaction-with-details-and-hidden=until-found">https://html.spec.whatwg.org/#interaction-with-details-and-hidden=until-found</a>

Otherwise, without this change, when a descendant node of a closed
&lt;details&gt; element contains a match for the given find-in-page search
string, or when navigating to a fragment ID for a descendant element of
a closed &lt;details&gt; element, the closed &lt;details&gt; element isn’t opened
automatically (not auto-expanded).

Further, per the requirements in the HTML spec’s Rendering section at
<a href="https://html.spec.whatwg.org/#the-details-and-summary-elements">https://html.spec.whatwg.org/#the-details-and-summary-elements</a>:the-details-element-6,
when the second slot of a &lt;details&gt; element has no “open” attribute,
this change causes that slot to have its “style” attribute set to
“display: block; content-visibility: hidden”.

Otherwise, without this change, when the “open” attribute is removed
from a &lt;details&gt; element, its second slot is removed from the DOM.

This change also introduces the DetailsAutoExpandEnabled preference
(disabled by default), for controlling whether &lt;details&gt; elements get
auto-expanded under the conditions described above.

And this also adds a testRunner.findInPage(target, options) function —
WK2-only — for testing the browser UI find-in-page feature, by calling
into the same code path that gets executed when that feature is used.

* LayoutTests/editing/text-iterator/find-in-page-in-closed-details-expected.txt: Added.
* LayoutTests/editing/text-iterator/find-in-page-in-closed-details.html: Added.
* LayoutTests/editing/text-iterator/find-in-page-in-summary-of-closed-details-expected.txt: Added.
* LayoutTests/editing/text-iterator/find-in-page-in-summary-of-closed-details.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-details-element/auto-expand-details-element-fragment-expected.txt:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/webrtc/legacy-api-expected.txt:
* LayoutTests/webrtc/legacy-api.html:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::expandAllDetailsAncestors):
* Source/WebCore/dom/Node.h:
* Source/WebCore/html/HTMLDetailsElement.cpp:
(WebCore::HTMLDetailsElement::attributeChanged):
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::scrollToFragmentInternal):
* Source/WebCore/page/LocalFrameViewLayoutContext.cpp:
(WebCore::LocalFrameViewLayoutContext::isSkippedContentForLayout const):
* Source/WebKit/WebProcess/WebPage/FindController.cpp:
(WebKit::FindController::findString):
* Tools/WebKitTestRunner/InjectedBundle/Bindings/TestRunner.idl:
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
(WTR::TestRunner::findStringMatchesInPage):
(WTR::TestRunner::findInPage):
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.h:
* Tools/WebKitTestRunner/TestInvocation.cpp:
(WTR::TestInvocation::didReceiveMessageFromInjectedBundle):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2118f670fe3b5d958686ed7f572dd24ce697f6c3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/90582 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/10114 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/45520 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/95594 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/41364 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/92635 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/10511 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/18433 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/69702 "Found 1 new test failure: imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-details-element/auto-expand-details-element-fragment.html (failure)") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27258 "Found 1 new test failure: http/tests/security/cross-origin-blob-transfer.html (failure)") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/93583 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/8013 "Found 2 new test failures: editing/text-iterator/find-in-page-in-closed-details.html http/tests/security/cross-origin-blob-transfer.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/82148 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50054 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/7750 "Found 1 new test failure: imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-details-element/auto-expand-details-element-fragment.html (failure)") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/36519 "Found 2 new test failures: http/tests/security/cross-origin-blob-transfer.html imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-details-element/auto-expand-details-element-fragment.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/40494 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/83392 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/78069 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/37587 "Found 3 new test failures: editing/text-iterator/find-in-page-in-closed-details.html http/tests/security/cross-origin-blob-transfer.html imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-details-element/auto-expand-details-element-fragment.html (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/97422 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/89362 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/17774 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/13079 "Found 3 new test failures: editing/text-iterator/find-in-page-in-closed-details.html http/tests/security/cross-origin-blob-transfer.html imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-details-element/auto-expand-details-element-fragment.html (failure)") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/78727 "Found 3 new test failures: editing/text-iterator/find-in-page-in-closed-details.html http/tests/security/cross-origin-blob-transfer.html imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-details-element/auto-expand-details-element-fragment.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/18032 "Built successfully") | [❌ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/77973 "Exiting early after 10 failures. 118 tests run. 1 failures") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/77927 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22365 "Passed tests") | [❌ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/20988 "Found 3 new test failures: editing/text-iterator/find-in-page-in-closed-details.html http/tests/security/cross-origin-blob-transfer.html imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-details-element/auto-expand-details-element-fragment.html (failure)") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/11096 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/17784 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/23116 "Built successfully") | [❌ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/111851 "Failed to checkout and rebase branch from PR 18785") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/17523 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/35/builds/111851 "Failed to checkout and rebase branch from PR 18785") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/20978 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/19307 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -2389,8 +2393,19 @@ bool LocalFrameView::scrollToFragmentInternal(StringView fragmentIdentifier)
+                    break;
+                }
+            }
+        }
```
This code is duplicated with FindController.cpp, can we make a common method on Element or such ?

sideshowbarker replied:
> This code is duplicated with FindController.cpp, can we make a common method on Element or such ?

Yes — made it so (though made it on Node rather than on Element, because the code where the caller is in FindController.cpp operates on a Node, not an element)

nt1m commented on Source/WebCore/dom/Node.cpp:
```diff
@@ -2102,6 +2106,21 @@ Element* Node::enclosingLinkEventParentOrSelf()
+        if (is<HTMLSummaryElement>(element))
+            continue;
+        if (is<HTMLDetailsElement>(element) && !element.hasAttributeWithoutSynchronization(HTMLNames::openAttr)) {
+            downcast<HTMLDetailsElement>(element).toggleOpen();
```
```suggestion
        if (RefPtr detailsElement = dynamicDowncast<HTMLDetailsElement>(element)) {
            if (!element.hasAttributeWithoutSynchronization(HTMLNames::openAttr)) {
                detailsElement->toggleOpen();
```

Can you use dynamicDowncast?

sideshowbarker replied:
> Can you use dynamicDowncast?

Yup, thanks — made it so

nt1m commented on Source/WebCore/html/HTMLDetailsElement.cpp:
```diff
@@ -155,15 +160,27 @@ void HTMLDetailsElement::attributeChanged(const QualifiedName& name, const AtomS
+                if (document().settings().detailsAutoExpandEnabled()) {
+                    m_defaultSlot->setInlineStyleProperty(CSSPropertyContentVisibility, CSSValueHidden);
+                    m_defaultSlot->setInlineStyleProperty(CSSPropertyDisplay, CSSValueBlock);
+                    if (auto* renderer = this->renderer())
```
```suggestion
                    if (RefPtr renderer = this->renderer())
```

cdumez replied:
renderer is not ref-counted. So what you actually want is:
```
if (CheckedPtr renderer = this->renderer())
```

sideshowbarker replied:
Thanks both — changed to using `CheckedPtr`

cdumez commented on Source/WebCore/dom/Node.cpp:
```diff
@@ -2102,6 +2106,21 @@ Element* Node::enclosingLinkEventParentOrSelf()
+        if (is<HTMLSummaryElement>(element))
+            continue;
+        if (is<HTMLDetailsElement>(element) && !element.hasAttributeWithoutSynchronization(HTMLNames::openAttr)) {
+            downcast<HTMLDetailsElement>(element).toggleOpen();
```
We should be holding a Ref / RefPtr to the element before calling toggleOpen() on it.

cdumez replied:
I see that Tim's proposal to switch to dynamicDowncast<> also addresses my concern too. Never mind.

annevk commented on LayoutTests/imported/w3c/web-platform-tests/css/css-animations/animation-canceled-by-parent-details-element-being-closed-expected.txt:
```diff
@@ -1,3 +1,5 @@
-PASS A CSS Animation running on an element within a <details> element is canceled after the <details> element is closed.
+Harness Error (TIMEOUT), message = null
+
+TIMEOUT A CSS Animation running on an element within a <details> element is canceled after the <details> element is closed. Test timed out
```
Per https://wpt.fyi/results/css/css-animations/animation-canceled-by-parent-details-element-being-closed.html it seems this is also the case for Chromium. I suggest we skip this test instead therefore (and maybe file an upstream issue to get it adjusted/removed).

annevk commented on Source/WebCore/dom/Node.cpp:
```diff
@@ -2102,6 +2106,23 @@ Element* Node::enclosingLinkEventParentOrSelf()
+        if (RefPtr detailsElement = dynamicDowncast<HTMLDetailsElement>(element)) {
+            if (!detailsElement->hasAttributeWithoutSynchronization(HTMLNames::openAttr)) {
+                detailsElement->toggleOpen();
+                break;
```
Why does this break here? The specification seems to suggest it should expand all ancestors, not just the first with a non-matching `open` attribute.

annevk reviewed (commented):
Overall this looks reasonable, but not sure it does the right thing when there's more than one closed ancestor.

I guess you plan to add the event handling in a follow-up change?

(Disclosure: I'm not a reviewer, but Tim asked if I could take a look.)

nt1m reviewed (commented):
Can we split this in two distinct changes? One that uses content-visibility/display (regardless of the settings flag), and this one for the auto-expand feature.

sideshowbarker committed:
Auto-expand “details” element for find-in-page and scrolling to fragments
https://bugs.webkit.org/show_bug.cgi?id=228843
rdar://81856620

Reviewed by NOBODY (OOPS!).

This change causes any closed <details> element to be opened
automatically (auto-expanded) if either:

- the user is using find-in-page and a descendant node of the closed
  <details> element contains a match for the given search string
- the user is navigating to a fragment ID and the fragment ID is for a
  descendant element of the closed <details> element

…per the requirements in the HTML spec’s Find-in-page section at
https://html.spec.whatwg.org/#interaction-with-details-and-hidden=until-found

Otherwise, without this change, when a descendant node of a closed
<details> element contains a match for the given find-in-page search
string, or when navigating to a fragment ID for a descendant element of
a closed <details> element, the closed <details> element isn’t opened
automatically (not auto-expanded).

Further, per the requirements in the HTML spec’s Rendering section at
https://html.spec.whatwg.org/#the-details-and-summary-elements:the-details-element-6,
when the second slot of a <details> element has no “open” attribute,
this change causes that slot to have its “style” attribute set to
“display: block; content-visibility: hidden”.

Otherwise, without this change, when the “open” attribute is removed
from a <details> element, its second slot is removed from the DOM.

This change also introduces the DetailsAutoExpandEnabled preference
(disabled by default), for controlling whether <details> elements get
auto-expanded under the conditions described above.

And this also adds a testRunner.findInPage(target, options) function —
WK2-only — for testing the browser UI find-in-page feature, by calling
into the same code path that gets executed when that feature is used.

* LayoutTests/editing/text-iterator/find-in-page-in-closed-details-expected.txt: Added.
* LayoutTests/editing/text-iterator/find-in-page-in-closed-details.html: Added.
* LayoutTests/editing/text-iterator/find-in-page-in-summary-of-closed-details-expected.txt: Added.
* LayoutTests/editing/text-iterator/find-in-page-in-summary-of-closed-details.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-details-element/auto-expand-details-element-fragment-expected.txt:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/webrtc/legacy-api-expected.txt:
* LayoutTests/webrtc/legacy-api.html:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::expandAllDetailsAncestors):
* Source/WebCore/dom/Node.h:
* Source/WebCore/html/HTMLDetailsElement.cpp:
(WebCore::HTMLDetailsElement::attributeChanged):
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::scrollToFragmentInternal):
* Source/WebCore/page/LocalFrameViewLayoutContext.cpp:
(WebCore::LocalFrameViewLayoutContext::isSkippedContentForLayout const):
* Source/WebKit/WebProcess/WebPage/FindController.cpp:
(WebKit::FindController::findString):
* Tools/WebKitTestRunner/InjectedBundle/Bindings/TestRunner.idl:
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
(WTR::TestRunner::findStringMatchesInPage):
(WTR::TestRunner::findInPage):
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.h:
* Tools/WebKitTestRunner/TestInvocation.cpp:
(WTR::TestInvocation::didReceiveMessageFromInjectedBundle):
